system/lib/pthread/pthread_create.c

```diff
@@ -232,6 +263,13 @@ void _emscripten_thread_free_data(pthread_t t) {
   }
 #endif
 
+  __tl_lock();
+
+  t->next->prev = t->prev;
+  t->prev->next = t->next;
```

Comment:
Why is this needed in addition to the block in _emscripten_thread_exit?  All threads should call _emscripten_thread_exit shouldn't they?

Reply:
There was a race condition, and that block was supposed to be a fix for it.

What was happening was that, inside pthread_create, the new pthread gets added to the circular list *after* the call to __pthread_create_js. This leads to a race where the new thread starts running and exits (i.e. calls _emscripten_thread_exit), before it gets added to the list inside pthread_create, meaning the list now contains a DT_EXITED thread that should not be there. 

The proper fix here is for the list to be updated *before* calling __pthread_create_js. If the call fails, the new pthread is then removed from the list. I have made that change.

However, I think it may still make sense to keep the block above, as there are cases where a pthread structure is freed using _emscripten_thread_free_data, without having called _emscripten_thread_exit:
i) In exitRuntime, terminateAllThreads is called which returns all workers to the thread pool and calls _emscripten_thread_free_data on them, regardless of whether they have exited. 
ii) Any thread that calls exit() *does not* call _emscripten_thread_exit either (see e=='unwind' in worker.js), which may be potentially problematic when EXIT_RUNTIME is not enabled.  Notably, in the PROXY_TO_PTHREAD case, the "main" thread is itself a pthread that ends up calling exit(), meaning it won't perform _emscripten_thread_exit cleanup either.

Given the above, it seems more correct to have that block above in _emscripten_thread_free_data.

Anyway, the above is just my beginner's understanding of the code, eager to know what you think!

Reply:
If a thread doesn't exit with `_emscripten_thread_exit` that would be a bug, and I don't think we should try to work around it unless we have to.

I suggest we leave `_emscripten_thread_free_data` alone, unless that change is needs to make a particular test pass?